homeassistant/components/idasen_desk/__init__.py

```diff
@@ -28,49 +29,92 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class IdasenDeskCoordinator(DataUpdateCoordinator):
+    """Class to manage updates for the Idasen Desk."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        name: str,
+        address: str,
+    ) -> None:
+        """Init IdasenDeskCoordinator."""
+
+        super().__init__(hass, logger, name=name)
+        self._address = address
+        self._expected_connected = False
+
+        self.desk = Desk(self.async_set_updated_data)
+
+    async def async_connect(self) -> bool:
+        """Connect to desk."""
+        _LOGGER.debug("Trying to connect %s", self._address)
+        ble_device = bluetooth.async_ble_device_from_address(
+            self.hass, self._address, connectable=True
+        )
+        if ble_device is None:
+            return False
+        self._expected_connected = True
+        await self.desk.connect(ble_device)
+        return True
+
+    async def async_disconnect(self) -> None:
+        """Disconnect from desk."""
+        _LOGGER.debug("Disconnecting from %s", self._address)
+        self._expected_connected = False
+        await self.desk.disconnect()
+
+    @callback
+    def async_set_updated_data(self, data: int | None) -> None:
+        """Handle data update."""
+        if self._expected_connected:
+            if not self.desk.is_connected:
+                _LOGGER.debug("Desk disconnected. Reconnecting")
+                self.hass.async_create_task(self.async_connect())
+        elif self.desk.is_connected:
+            _LOGGER.warning("Desk is connected but should not be. Disconnecting")
+            self.hass.async_create_task(self.desk.disconnect())
+        return super().async_set_updated_data(data)
+
+
 @dataclass
 class DeskData:
     """Data for the Idasen Desk integration."""
 
-    desk: Desk
     address: str
     device_info: DeviceInfo
-    coordinator: DataUpdateCoordinator
+    coordinator: IdasenDeskCoordinator
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up IKEA Idasen from a config entry."""
     address: str = entry.data[CONF_ADDRESS].upper()
 
-    coordinator: DataUpdateCoordinator = DataUpdateCoordinator(
-        hass,
-        _LOGGER,
-        name=entry.title,
+    coordinator: IdasenDeskCoordinator = IdasenDeskCoordinator(
```

Comment:
We're missing a type annotation for the data type of the coordinator. The coordinator class is generic. See other integrations that implement a custom coordinator class for examples.